File: Haskell/Implementations/Abstract Factory/Knife.hs
module Knife
    ( Knife
    , knifeWeight
    , knifeDoubleEdged
    , cutFood
    , mkPrimitiveKnife
    , mkSilverwareKnife
    ) where

data Knife = Knife
    { knifeWeight      :: Int
    , knifeDoubleEdged :: Bool
    , cutFood          :: String -> IO ()
    }

mkPrimitiveKnife :: Knife
mkPrimitiveKnife = Knife
    { knifeWeight      = 110
    , knifeDoubleEdged = True
    , cutFood          = \food -> putStrLn
        (  "You attempt to cut the "
        ++ food
        ++ ", but you instead end up clumsily crushing it."
        )
    }

mkSilverwareKnife :: Knife
mkSilverwareKnife = Knife
    { knifeWeight      = 240
    , knifeDoubleEdged = False
    , cutFood          = \food ->
        putStrLn ("The " ++ food ++ " slices easily, and with finesse.")
    }

File: Java/Implementations/Abstract Factory/Knife.java
public abstract class Knife {
	protected boolean isDoubleEdged;
	protected int weight;

	protected Knife(boolean isDoubleEdged, int weight) {
		this.isDoubleEdged = isDoubleEdged;
		this.weight = weight;
	}

	public boolean isDoubleEdged() {
		return this.isDoubleEdged;
	}

	public int weight() {
		return this.weight;
	}

	public abstract void cutFood(String food);
}
